Registry: Adds doc on  aliases (including protected aliases) (#1255)
wandb/docs: content/guides/core/registry/aliases.md

{{% alert title="When to use a tag versus using an alias" %}}
Use aliases to reference a specific artifact version. Each alias within a collection is unique. Only one artifact version can have a specific alias at a time.

Use tags to organize and group artifact versions or collections based on a common theme. Multiple artifact versions and collections can share the same tag.
{{% /alert %}}

## Default aliases

W&B automatically assigns the following aliases to each artifact version you link with the same name:

* The `latest` alias to the most recent artifact version you link to a collection.
* A unique version number. W&B counts each artifact version (zero indexing) you link. W&B uses the count number to assign a unique version number to that artifact.

For example, if you link an artifact named `zoo_model` three times, W&B creates three aliases `v0`, `v1`', and `v2` respectively.

> ## Custom aliases

Reviewer: can we link to the automation docs as well and mention that aliases (or adding an alias) is also used as a triggering event?

The following code snippet demonstrates how to link an artifact version to a collection and add aliases to that artifact version with the Python SDK. Replace values within `<>` with your own:

```python
import wandb

# Initialize a run
run = wandb.init(entity = "<team_entity>", project = "<project_name>")

# Create an artifact object
# The type parameter specifies both the type of the 
# artifact object and the collection type
artifact = wandb.Artifact(name = "<name>", type = "<type>")

# Add the file to the artifact object. 
# Specify the path to the file on your local machine.
artifact.add_file(local_path = "<local_path_to_artifact>")

# Specify the collection and registry to link the artifact to
REGISTRY_NAME = "<registry_name>"
COLLECTION_NAME = "<collection_name>"
target_path=f"wandb-registry-{REGISTRY_NAME}/{COLLECTION_NAME}"

# Link the artifact version to the collection
# Add one or more aliases to this artifact version
run.link_artifact(
    artifact = artifact, 
    target_path = target_path, 
    aliases = ["<alias_1>", "<alias_2>"]
    )
```
{{% /tab %}}
{{< /tabpane >}}

> ### Protected aliases

Reviewer: should we consider having this be separate section on lifecycle management and permissions? Rather than stick it into alias section?

Author: I don't think so. An alias, protected or not, fall into the same bucket of pointing to specific artifact version. It also flows with the overall workflow of our current table of contents. (As a new user I'd expect to see protect aliases in the same page where I learn about general aliases). 

This page is also relatively short with only two sections so it won't get lost.

> Use a [protected alias]({{< relref "aliases/#protected-aliases" >}}) to both label and identify artifact versions that should not be modified or deleted. For example, consider using a `production` protected alias to label and identify artifact versions that are in used in your organization's machine learning production pipeline.

Reviewer: users also cannot delete a collection that contains a version with a protected alias, they also cannot unlink it from the registry

Only [registry admins]({{< relref "/guides/core/registry/configure_registry/#registry-roles" >}}) can create protected aliases and add or remove protected aliases from an artifact version. See [Configure registry access]({{< relref "/guides/core/registry/configure_registry.md" >}}) for information on how to manage users and assign roles in a registry.

<!-- Admins can only create protected aliases in the W&B Registry UI. Once created, an admin can add a protected alias to an artifact version with the W&B Registry UI or the Python SDK. -->

<!-- Protected aliases are not shared across registries.  -->

Common protected aliases include:

- **Production**: The artifact version is ready for production use.
- **Staging**: The artifact version is ready for testing.

> #### Create protected aliases

Reviewer: can we add a gif for this workflow?

Author: Unless critical, we avoid gifs. It's a technical debt that quickly becomes outdated.

The following steps describe how to create a protected alias in the W&B Registry UI:

1. Navigate to the Registry App.
2. Select a registry.
3. Click the gear button on the top right of the page to view the registry's settings.
4. Within the **Protected Aliases** section, click on the plus icon (**+**) to add one or more protected aliases.

If successful, you will see one or more protected aliases appear as a gray rectangle with a lock icon in the **Protected Aliases** section. 

{{% alert %}}
Unlike custom non protected aliases, protected aliases can only be created in the W&B Registry UI and not programmatically with the Python SDK. Once created, admins can add a protected alias to an artifact version with the W&B Registry UI or the Python SDK.
{{% /alert %}}

The following steps describe how to add a protected alias to an artifact version with the W&B Registry UI:

1. Navigate to the W&B Registry.
2. Click the **View details** button in a collection.
3. Within the **Versions** section, select the **View** button for a specific artifact version.
4. Select the plus icon (**+**) to add one or more protected aliases next to the **Aliases** field.

> Admins can add a protected alias to an artifact version programmatically with the Python SDK. Make sure that the protected alias already exists in the registry.

Reviewer: helpful to add a code snippet or at least hyperlink to the above section

---

> title: Use aliases to point to a specific artifact version

Reviewer 2: This is a pretty long title and seems to have plural/singular disagreement with "aliases" and "a specific artifact version". I glean from the intro that a single version can have more than one alias, and maybe that's what this is trying to convey. I like "reference" a little more than "point to".

Author: Not a fan of my initial title either. 

An artifact version can have one or more aliases; a combo of protected and non protected aliases.

weight: 5
---

> Reference a specific artifact version with an alias. An artifact version can have one or more aliases. [W&B automatically assigns aliases]({{< relref "aliases#default-aliases" >}}) to each artifact you link with the same name. You can also [create one or more custom aliases]({{< relref "aliases#custom-aliases" >}}) to reference a specific artifact version.

Reviewer 2: "artifact version" could be a link to a relevant Registry page. It's not immediately clear to me why I would want to use an alias. Is it because default artifact names are unwieldy?

Author: I added the info blurb below this section to help clarify this point. Was that at all useful or nay? If nay, I'll happily had more context. :)

Aliases appear as rectangles with the name of that alias in the rectangle in the Registry UI. If an [alias is protected]({{< relref "aliases#protected-aliases" >}}), it appears as a gray rectangle with a lock icon. Otherwise, the alias appears as an orange rectangle. Aliases are not shared across registries.

> {{% alert title="When to use a tag versus using an alias" %}}

Reviewer 2: ```suggestion
{{% alert title="When to use a tag or an alias" %}}
```
The title has tags first but the admonition talks about aliases first. Worth harmonizing?

> Use aliases to reference a specific artifact version. Each alias within a collection is unique. Only one artifact version can have a specific alias at a time.

Reviewer 2: ```suggestion
Use an alias to reference a specific artifact version. Each alias within a collection is unique. Only one artifact version can have a specific alias at a time.
```
Why would you assign more than one alias to a version?

Author: Depends on the workflow. For example, someone might have multiple aliases for triggering different automations using the same alias.  Maybe a "prod" and a "test-model-prod", etc.

Author: The % of folks who do this, I do not know. But we offer it for flexibility.

Use tags to organize and group artifact versions or collections based on a common theme. Multiple artifact versions and collections can share the same tag.
{{% /alert %}}

> 

Reviewer 2: ```suggestion
```

> When you add an alias to an artifact version, you can optionally trigger a workflow step such a Slack notification or webhook. For more information, see the [Registry section]({{< relref  "/guides/core/automations/automation-events/#registry" >}}) in [Automation event and scopes]({{< relref "/guides/core/automations/automation-events/" >}}).

Reviewer 2: ```suggestion
When you add an alias to an artifact version, you can optionally start a [Registry automation]({{< relref  "/guides/core/automations/automation-events/#registry" >}}) to notify a Slack channel or trigger a webhook.
```

## Default aliases

W&B automatically assigns the following aliases to each artifact version you link with the same name:

* The `latest` alias to the most recent artifact version you link to a collection.
* A unique version number. W&B counts each artifact version (zero indexing) you link. W&B uses the count number to assign a unique version number to that artifact.

> For example, if you link an artifact named `zoo_model` three times, W&B creates three aliases `v0`, `v1`', and `v2` respectively.

Reviewer 2: ```suggestion
For example, if you link an artifact named `zoo_model` three times, W&B creates three aliases `v0`, `v1`', and `v2` respectively. `v2` also has the `latest` alias.
```

> For example, if you link an artifact named `zoo_model` three times, W&B creates three aliases `v0`, `v1`', and `v2` respectively.

Reviewer 3: tiny nit:
```suggestion
For example, if you link an artifact named `zoo_model` three times, W&B creates three aliases `v0`, `v1`, and `v2` respectively.
```

## Custom aliases

> Create custom aliases to reference a specific artifact versions based on your unique use case. As an example, you might use aliases `dataset_version_v0`, `dataset_version_v1`, `dataset_version_v2`, and so forth to identify which dataset a model was trained on. As an another example, you might use a `best_model` alias to keep track of the best performing artifact model version.

Reviewer 2: ```suggestion
Create one or more custom aliases for a specific artifact versions based on your unique use case. For example:

- You might use aliases such as `dataset_version_v0`, `dataset_version_v1`, and `dataset_version_v2` to identify which dataset a model was trained on.
- You might use a `best_model` alias to keep track of the best performing artifact model version.
```

> Any user with a [Member or Admin registry role]({{< relref "guides/core/registry/configure_registry/#registry-roles" >}}) can add or remove a custom alias from a linked artifact. Consider using a [protected alias]({{< relref "aliases/#protected-aliases" >}}) to label and identify artifact versions that should not be modified or deleted.

Reviewer 2: ```suggestion
Any user with a [Member or Admin role]({{< relref "guides/core/registry/configure_registry/#registry-roles" >}}) on a registry can add or remove a custom alias from a linked artifact in that registry. If appropriate, use [protected aliases]({{< relref "aliases/#protected-aliases" >}}) to label and identify which artifact versions to protect from modification or deletion.
```

Author: In my latest commit I left "role" in " Member or Admin registry role" to help differentiate registry roles from org roles. (They are different).

Reviewer 2: Thanks, I like it. I'd prefer something more active than "should not be modified or deleted."

> You can create a custom alias with the W&B Registry or the Python SDK. Based on your use case, click on a tab below that best fits your needs.

Reviewer 2: ```suggestion
You can create a custom alias with the W&B Registry App or the Python SDK. Based on your use case, click on a tab below that best fits your needs.
```

{{< tabpane text=true >}}
{{% tab header="W&B Registry" value="app" %}}

1. Navigate to the W&B Registry.
2. Click the **View details** button in a collection.

> 3. Within the **Versions** section, select the **View** button for a specific artifact version.

Reviewer 2: ```suggestion
3. Within the **Versions** section, click the **View** button for a specific artifact version.
```

> 4. Select the plus icon (**+**) to add one or more aliases next to the **Aliases** field.

Reviewer 2: ```suggestion
4. Click (**+**) to add one or more aliases next to the **Aliases** field.
```
If the button text is actually "+", this is accessible.

{{% /tab %}}

{{% tab header="Python SDK" value="python" %}}

> When you link an artifact version to a collection with the Python SDK you can optionally provide a list of one or more aliases as an argument to the `alias` parameter in [`link_artifact()`]({{< relref "/ref/python/run/#link_artifact" >}}). W&B creates an alias for you if the alias you provide does not already exist.

Reviewer 2: ```suggestion
When you link an artifact version to a collection with the Python SDK you can optionally set the `alias` parameter in  [`link_artifact()`]({{< relref "/ref/python/run/#link_artifact" >}}) to a list of one or more aliases. If necessary, W&B creates the alias before assigning it.
```
I had to read this a couple times in comparison with line 108. Is it worth calling out here that protected aliasee don't have this behavior of creating the alias automatically?

The following code snippet demonstrates how to link an artifact version to a collection and add aliases to that artifact version with the Python SDK. Replace values within `<>` with your own:

```python
import wandb

# Initialize a run
run = wandb.init(entity = "<team_entity>", project = "<project_name>")

# Create an artifact object
# The type parameter specifies both the type of the 
# artifact object and the collection type
artifact = wandb.Artifact(name = "<name>", type = "<type>")

# Add the file to the artifact object. 
# Specify the path to the file on your local machine.
artifact.add_file(local_path = "<local_path_to_artifact>")

# Specify the collection and registry to link the artifact to
REGISTRY_NAME = "<registry_name>"
COLLECTION_NAME = "<collection_name>"
target_path=f"wandb-registry-{REGISTRY_NAME}/{COLLECTION_NAME}"

# Link the artifact version to the collection
# Add one or more aliases to this artifact version
run.link_artifact(
    artifact = artifact, 
    target_path = target_path, 

>     aliases = ["<alias_1>", "<alias_2>"]

Reviewer 2: It might be beneficial to use more realistic placeholders to help the reader imagine what they might use it for.

    )
```
{{% /tab %}}
{{< /tabpane >}}

### Protected aliases
Use a [protected alias]({{< relref "aliases/#protected-aliases" >}}) to both label and identify artifact versions that should not be modified or deleted. For example, consider using a `production` protected alias to label and identify artifact versions that are in used in your organization's machine learning production pipeline.

> Only [registry admins]({{< relref "/guides/core/registry/configure_registry/#registry-roles" >}}) can create protected aliases and add or remove protected aliases from an artifact version. Users (Member and Viewer roles) can not delete a collection that contains a version with a protected alias, and they also cannot unlink it from the registry. See [Configure registry access]({{< relref "/guides/core/registry/configure_registry.md" >}}) for information on how to manage users and assign roles in a registry.

Reviewer 2: ```suggestion
Only [registry admins]({{< relref "/guides/core/registry/configure_registry/#registry-roles" >}}) can create protected aliases and add or remove protected aliases from an artifact version. Members and Viewers cannot unlink a protected version or delete a collection that contains a protected . See [Configure registry access]({{< relref "/guides/core/registry/configure_registry.md" >}}) for details.
```

<!-- Admins can only create protected aliases in the W&B Registry UI. Once created, an admin can add a protected alias to an artifact version with the W&B Registry UI or the Python SDK. -->

> <!-- Protected aliases are not shared across registries.  -->

Reviewer 2: Intentionally commented?

Common protected aliases include:

- **Production**: The artifact version is ready for production use.

> - **Staging**: The artifact version is ready for testing.

Reviewer 2: It might be good to present these as tuples and to have another example or two:

- **Production** and **Staging** to indicate that artifact version's lifecycle.
- <something else>

> #### Create protected aliases

Reviewer 2: ```suggestion
#### Create a protected alias
```

The following steps describe how to create a protected alias in the W&B Registry UI:

1. Navigate to the Registry App.
2. Select a registry.
3. Click the gear button on the top right of the page to view the registry's settings.

> 4. Within the **Protected Aliases** section, click on the plus icon (**+**) to add one or more protected aliases.

Reviewer 2: Similar feedback as above about the + button.

> If successful, you will see one or more protected aliases appear as a gray rectangle with a lock icon in the **Protected Aliases** section. 

Reviewer 2: ```suggestion
After creation, each protected alias appears as a gray rectangle with a lock icon in the **Protected Aliases** section. 
```

{{% alert %}}

> Unlike custom non protected aliases, protected aliases can only be created in the W&B Registry UI and not programmatically with the Python SDK. Once created, admins can add a protected alias to an artifact version with the W&B Registry UI or the Python SDK.

Reviewer 2: ```suggestion
Unlike custom aliases that are not protected, creating protected aliases is available exclusively in the W&B Registry UI and not programmatically with the Python SDK. To add a protected alias to an artifact version, you can use the W&B Registry UI or the Python SDK.
```

{{% /alert %}}

The following steps describe how to add a protected alias to an artifact version with the W&B Registry UI:

1. Navigate to the W&B Registry.
2. Click the **View details** button in a collection.
3. Within the **Versions** section, select the **View** button for a specific artifact version.
4. Select the plus icon (**+**) to add one or more protected aliases next to the **Aliases** field.

> Admins can add a protected alias to an artifact version programmatically with the Python SDK. Make sure that the protected alias already exists in the registry.

Reviewer 2: Only one admin can add a protected alias, right? There are a few other instances of subject/object count disagreement.

Author: A registry can have multiple admins.

Reviewer 2: Yeah understood, but only one admin at a time can add a given alias at a time. So maybe "Any admin can"? I can see either way, tbh!

> The following code snippet demonstrates how to both link an artifact version to a collection and add aliases to that artifact version with the Python SDK. Replace values within `<>` with your own:

Reviewer 2: I wonder if it is worth moving both of the example code blocks to an ## Examples section down here. I felt like I fell off the end of this page with no Next Steps or anything.

Author: I removed this code block and instead moved it up. It's the same code block so it's a bit redundant.
